builder.nix

```diff
@@ -106,28 +117,32 @@ pkgs.stdenv.mkDerivation {
     bash perl python3 dtc
   ] ++ lib.optional (!lib.versionAtLeast release "21" && release != "snapshot") python2;
 
-  buildPhase = let
-    makeArgs = lib.concatStringsSep " " [
-      "-j$NIX_BUILD_CORES"
-      "DISABLED_SERVICES='${lib.concatStringsSep " " disabledServices}'"
-      "PACKAGES='${lib.concatStringsSep " " packages}'"
-      "PROFILE='${profile}'"
-      "SHELL='${pkgs.runtimeShell}'"
-      (lib.optionalString (files != null) "FILES='./files'")
-      (lib.optionalString (rootFsPartSize != null) "ROOTFS_PARTSIZE=${toString rootFsPartSize}")
-      (lib.optionalString (extraImageName != null) "EXTRA_IMAGE_NAME='${extraImageName}'")
-    ];
-
-  in lib.optionalString (lib.versionOlder release "19") ''
-    # Hack around broken check for gcc
-    touch staging_dir/host/.prereq-build
-  '' + lib.optionalString (files != null) ''
-      # copy files to avoid making etc read-only
-      cp -r --no-preserve=all ${files} files
-  '' + ''
-    make image ${makeArgs}
-  '';
 
+  buildFlags = [
+    ".DEFAULT_GOAL=image"
+  ];
+
+  preBuild =
+    let
+      mkBuildFlagsArray = flags: lib.concatStringsSep " " (lib.mapAttrsToList (name: value: ''${name}="${builtins.toString value}"'') flags);
+
+      flags = {
+        PROFILE = profile;
+        PACKAGES = lib.concatStringsSep " " packages;
+      } // lib.optionalAttrs (disabledServices != [ ]) {
+        DISABLED_SERVICES = lib.concatStringsSep " " disabledServices;
+      } // lib.optionalAttrs (files != null) {
+        FILES = "./files";
+      } // lib.optionalAttrs (rootFsPartSize != null) {
+        ROOTFS_PARTSIZE = rootFsPartSize;
+      } // lib.optionalAttrs (extraImageName != null) {
+        EXTRA_IMAGE_NAME = extraImageName;
+      };
+    in
+    ''
+      buildFlagsArray+=(${mkBuildFlagsArray flags}) 
```

Comment:
I am using the `buildFlagsArray` here because those flags have spaces in them and `buildFlags`/`makeFlags` variables [don't support values with spaces](https://nixos.org/manual/nixpkgs/stable/#var-stdenv-makeFlagsArray) in them 😢

Reply:
I am using the `buildFlagsArray` here because those flags have spaces in them and `buildFlags`/`makeFlags` variables [don't support values with spaces](https://nixos.org/manual/nixpkgs/stable/#var-stdenv-makeFlagsArray) in them 😢